Unconfirmed tx address list

papacarp opened the pull request:
This PR addresses the unconfirmed transactions not visible in address view and input amounts not visible in transaction view.

This should close #277 , #130 
This could close #142 with a few more small changes - it should be easy to add those from mempool. 

There were two major areas of work:
1.  Blending unconfirmed transactions with confirmed transactions in the current address view.  Mostly extending what @chappjc  already started.
2.  Compensating for AmountIn values not set in dcrd calls to mempool transactions.

Known Issues
1.  Unconfirmed transactions are added at the start of the address view list regardless of indexing (start/count) settings.  I left some commented code in there for now that does the final slice to illustrate where we would do it, but we can remove before merge.  Fixing the indexing issue is a broader issue than this PR and is likely going to happen with PR#427 or I can fix it separately with a variation on the experimental PR I proposed (PR #433 but I do not think we need to add the column to the addresses table anymore).
2.  DCRD is fixing the AmountIn issue.  Depending on the status of that change the repository owners can make a determination if we should leave out those compensating changes or include in this PR.  

**Transaction View**

*Before*
![image](https://user-images.githubusercontent.com/11194546/40821251-88033c38-651a-11e8-8408-03513c2ceaaa.png)

*After*
![image](https://user-images.githubusercontent.com/11194546/40821266-94af5bf6-651a-11e8-9fed-89f4947583d5.png)

**Address View**

*Before*
![image](https://user-images.githubusercontent.com/11194546/40821292-cc1d8360-651a-11e8-960b-c8aae732626f.png)

*After*
![image](https://user-images.githubusercontent.com/11194546/40821297-d2c057ce-651a-11e8-84f9-c00485dc5178.png)

*Before*
![image](https://user-images.githubusercontent.com/11194546/40821332-18b96eaa-651b-11e8-89ab-664d0a025dda.png)

*After*
![image](https://user-images.githubusercontent.com/11194546/40821337-2335c75c-651b-11e8-8a1c-257a40582074.png)
Note:  age is updated

*Before*
![image](https://user-images.githubusercontent.com/11194546/40821384-77c9df88-651b-11e8-95bd-84d45f6b1e14.png)

*After*
![image](https://user-images.githubusercontent.com/11194546/40821400-87c3af9a-651b-11e8-90a8-3474bef5ebf6.png)
Note:  order of transactions is corrected.  Also note that unconfirmed tx is 1, since the transaction only has one confirmation.

*Before*
![image](https://user-images.githubusercontent.com/11194546/40821524-527e9c86-651c-11e8-91f9-6fc0c990b2bf.png)

*After*
![image](https://user-images.githubusercontent.com/11194546/40821529-59aef974-651c-11e8-81c9-9fab64fcfc66.png)

raedah commented:
I would not work around the dcrd issue unless important functionality is broken. Dcrdata is meant to accurately display what dcrd reports. If the work around does need to be used, https://github.com/decred/dcrdata/issues/130  and https://github.com/decred/dcrdata/issues/142 should be kept open, or a new issue created as a reminder to remove the temporary fix after the dcrd issue is resolved.

raedah commented:
Looks like tx could be subsorted by in[] or out[]

papacarp commented:
Rebased and updated to sort by in/out if time is same.  Note transaction confirmations are showing as different for the same tx.  Is that expected?

![image](https://user-images.githubusercontent.com/11194546/41242840-74ad9e38-6d55-11e8-9271-bb2d25c5ad4c.png)

This is ready for review @chappjc

chappjc commented:
Nope, definitely not expected.  Looks like it's showing confirms for the funding transaction rather than the spending transaction.

papacarp commented:
I can rebase this today or tomorrow.  The confirmation issue was not fixed AFAIK.  Do you want me to go ahead and fix issue #507 as well as part of this PR?

chappjc commented:
If you see the root cause and it is trivial to fix, yes, it's cool to include the fix here.  Otherwise this can go in and we can fix the confirmations issue in a follow-up.  Thanks!

papacarp commented:
rebased.  The  issue #507 seems to have fixed itself in the address view updates @dmigwi did.  However, i see two potentially new issues - I have confirmed they are in master right now.  I suspect both are related.

1.  unspent is error if the total unspent balance is less than zero.
2.  total unspent balance is likely less than zero because this address view has missing transactions.  There should be 30 transactions, but tip only pulls 20 transactions.

looks like something changed in "exp.explorerSource.AddressHistory' so it doesn't get the right balance.

current view: 20 transactions
![image](https://user-images.githubusercontent.com/11194546/43166515-9823541c-8f4b-11e8-9d44-42e40715fa03.png)

compare to beta:  https://beta.dcrdata.org/address/DsiJmwqFbsCEc1H3HF8UeRjoqJtKt9ZFRKw 

This PR can be merged into master once your review is complete.  I recommend we open up new bugs for the issues noted above.

Mike

dmigwi commented:
>unspent is error if the total unspent balance is less than zero.
This is a concern that we should address ASAP.

>total unspent balance is likely less than zero because this address view has missing transactions. There should be 30 transactions, but tip only pulls 20 transactions.

This concern is addressed by the links source and spent.  Notice the time in txs that appear here https://beta.dcrdata.org/address/DsiJmwqFbsCEc1H3HF8UeRjoqJtKt9ZFRKw. They are not well ordered but what appears in the above screenshot is well order. That was the main reason why we had to alter the addresses table.

chappjc commented:
@papacarp, @dmigwi please coordinate on those issues and their resolutions.  I'll get this PR in as is.
